MCO-1117: Make local MCO image builds faster

cheesesashimi opened the pull request:
**- What I did**
Go is very good about doing incremental builds, provided that it has a build cache that it can access. This is why once you've performed an initial `$ go build .`, future runs are much faster since it will only rebuild changes. Because of how our Dockerfile is written, it cannot take advantage of that caching. Each build takes place against an empty cache, which means that it must perform a full build every time. Recently, I learned that one can mount a cache into the build context at an arbitrary path by adding a  `--mount` flag to any `RUN` directive.

Doing this and using separate cache paths for the RHEL8 and RHEL9 builds in our Dockerfile, Podman can mount a cache directory at build-time that the Go toolchain can use to cache incremental build objects. Additionally, by moving the nmstate installation into a separate build stage, adding a package cache, and ensuring that dnf does not wipe out that package cache, that stage can be cached more effectively. This has the additional effect that once this stage is cached, one does not need a VPN connection for rebuilds.

It is worth mentioning that the primary benefit of this is only for local container image builds by MCO developers. While these changes will run without issue in CI, there is no likelihood of a cache hit in CI because the build contexts and their caches are destroyed between builds.

To back these changes up with some numbers:

Initial build:
```console
$ time podman build -t localhost/machine-config-operator:latest . 
430.41s user 115.37s system 347% cpu 2:36.85 total
```

Incremental rebuild without these Dockerfile changes (I added an env var lookup to `cmd/machine-config-operator/main.go` to force cache invalidation / rebuild):
```console
$ time podman build -t localhost/machine-config-operator:latest . 
432.58s user 115.05s system 357% cpu 2:33.15 total
```

Incremental build with Dockerfile changes (after doing an initial rebuild, I re-added the aforementioned changes).
```console
$ time podman build -t localhost/machine-config-operator:latest . 
49.96s user 33.70s system 169% cpu 49.275 total
```

**- How to verify it**
1. Clone this PR and run `$ time podman build -t localhost/machine-config-operator:latest .`.
2. Next, make a dummy change to the MCO code.
3. Re-run `$ time podman build -t localhost/machine-config-operator:latest .`.
4. The rebuild should be significantly faster.

**- Description for the changelog**
Make local MCO image builds faster

cheesesashimi commented:
/test test-unit

cheesesashimi commented:
/test unit

cheesesashimi commented on Dockerfile:
```diff
@@ -6,22 +6,27 @@ WORKDIR /go/src/github.com/openshift/machine-config-operator
 # FIXME once we can depend on a new enough host that supports globs for COPY,
 # just use that.  For now we work around this by copying a tarball.
-RUN make install DESTDIR=./instroot && tar -C instroot -cf instroot.tar .
+ENV GOCACHE="/go/rhel8/.cache"
```
I gave this and the RHEL9 stages separate caches so they don't step on each others toes and invalidate each other.

djoshy commented:
Works great for me in my local tests, lgtm from my side

djoshy commented:
/retest-required

djoshy commented:
/retest-required

dkhater-redhat commented:
this all looks good to me- just curious, how much time were you able to save with caching? i remember on my end it wasnt too time saving but was curious to see what you came up with

cheesesashimi commented:
If I'm doing an initial image build, there is no improvement. However, all subsequent builds can be done in ~50 seconds, or so, assuming that all I've changed is Go code and nothing in the `manifests/` directory. The big reason for that is because I moved nmstate installation into a separate build stage and instead of copying the whole MCO repo, I only copy the `manifests/` directory. This makes a cache hit much more likely and it's this cache hit where the biggest gain comes from.

djoshy commented:
/retest-required

cheesesashimi commented:
These failures might indicate there is something wrong with my changes to `Dockerfile.rhel7`, which I think is what builds our release image. I don't _have_ to make the changes there, but if someone were to run `$ make Dockerfile.rhel7`, those changes would get propagated.

cheesesashimi commented:
I think I figured it out. This needed to be rebased and I was unaware of that fact.

djoshy commented:
/lgtm
